### PR TITLE
Add memory optimization to transform dialect and enable it for reduction flow

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 1]
```diff
@@ -1104,5 +1115,58 @@ transform_dialect::IREEEraseHALDescriptorTypeFromMemRefOp::apply(
...
+  for (Operation *op : opToErase) {
+    op->erase();
+  }
+}
+
+DiagnosedSilenceableFailure
+transform_dialect::ApplyBufferOptimizationsOp::applyToOne(
+    Operation *target, transform::ApplyToEachResultList &results,
+    transform::TransformState &state) {
+  vector::transferOpflowOpt(target);
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:1155]: hmmm what is this?
Can you please add comments?

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td [hunk 2]
```diff
@@ -368,4 +371,35 @@ def ConfigExtractPart :
   }];
 }
 
+def ApplyBufferOptimizationsOp : Op<Transform_Dialect, "iree.apply_buffer_optimizations",
+    [TransformEachOpTrait,
+     TransformOpInterface,
+     DeclareOpInterfaceMethods<MemoryEffectsOpInterface>]> {
+  let description = [{
+    This applies memory optimization on memref. In particular it does store to
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td:379]: Doc and return modes need to be significantly beefed up.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 3]
```diff
@@ -184,7 +191,9 @@ static void addEraseUnnecessaryTensorOperandsPatterns(
 static void addRankReducingPatterns(RewritePatternSet &patterns) {
   populateReshapeToInterfaceTensorPatterns(patterns);
   vector::populateCastAwayVectorLeadingOneDimPatterns(patterns);
-  linalg::populateFoldUnitExtentDimsViaSlicesPatterns(patterns);
+  // TODO: Remove tensor unit dims causes problem of incompatibility between
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:194]: Is this covered by https://reviews.llvm.org/D141735 ?

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td [hunk 4]
```diff
@@ -47,6 +47,8 @@ def ApplyPatternsOp : Op<Transform_Dialect, "iree.apply_patterns",
       memref.subview.
       - fold_reassociative_reshapes: adds patterns that fold insert_slice/
       extract_slice ops with reassociative reshape ops.
+      - lower_transfer_op_permutations: Lower transfer ops to transfer ops
```
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td:50]: did you mean `lower_transfer_ops_to_identity_permutations`?

**Pull request closed without merging**